### ISSUE TITLE
mypy throws spurious errors when it cannot assertain `not None`

### ISSUE THREAD

**Issue by pkit** [python/mypy#15207]
**Bug Report**

Essentially reopening #4805 which was closed for no reason.

**To Reproduce**

```python
def check_something(func):
    def check(self, *args, **kwargs):
        if self.something is None:
            raise RuntimeError("Something is None")

        return func(self, *args, **kwargs)
    return check


class Something:
    def do1(self) -> None:
        print("do1")

    def do2(self) -> None:
        print("do2")


class MyClass(object):
    def __init__(self) -> None:
        self.something: Something | None = None

    @check_something
    def method1(self) -> None:
        self.something.do1()

    @check_something
    def method2(self) -> None:
        self.something.do2()

```

**Expected Behavior**

Totally valid code

**Actual Behavior**

```
$ python -m mypy test.py
test.py:24: error: Item "None" of "Optional[Something]" has no attribute "do1"  [union-attr]
test.py:28: error: Item "None" of "Optional[Something]" has no attribute "do2"  [union-attr]
Found 2 errors in 1 file (checked 1 source file)
```

**Your Environment**

- Mypy version used: mypy 1.2.0 (compiled: yes)
- Python version used: 3.10

**Thoughts**

Obviously the error can be fixed by doing `assert self.something is not None` in **every** method.
But that defeats the whole purpose of a decorator.
And what if there are 20 methods?
And what if there are 20 `self.somethingN` to check?
Writing bad copy-paste code just to please type checker?

**Comment by hauntsaninja** [comment]
Being rude on issues that are only tangentially related to yours is not an acceptable way of interacting with this community.
